pkg/handler/v2/external_cluster/kubeone.go

```
@@ -74,55 +79,165 @@ func importKubeOneCluster(ctx context.Context, name string, userInfoGetter func(
 	manifestSecret.Data = map[string][]byte{
-		resources.KubeOneManifest: yamlManifest,
+		resources.KubeOneManifest: patchManifest,
 	}
 	if err := masterClient.Patch(ctx, manifestSecret, ctrlruntimeclient.MergeFrom(oldManifestSecret)); err != nil {
-		return nil, fmt.Errorf("failed to update kubeone manifest secret for upgrade version %s/%s: %w", manifest.Name, manifest.Namespace, err)
+		return nil, fmt.Errorf("failed to update kubeone manifest secret for container-runtime containerd %s/%s: %w", manifest.Name, manifest.Namespace, err)
 	}
 
 	oldexternalCluster := externalCluster.DeepCopy()
 	// update kubeone externalcluster status.
 	externalCluster.Spec.CloudSpec.KubeOne.ClusterStatus.Status = kubermaticv1.StatusReconciling
-	if upgradeMD == "true" {
-		externalCluster.Spec.CloudSpec.KubeOne.ClusterStatus.StatusMessage = "upgrading machine MD"
-	}
-	// update api externalcluster status
-	newCluster.Status.State = apiv2.RECONCILING
+	// update api externalcluster status.
+	newCluster.Status = apiv2.ExternalClusterStatus{State: apiv2.RECONCILING}
+
 	if err := masterClient.Patch(ctx, externalCluster, ctrlruntimeclient.MergeFrom(oldexternalCluster)); err != nil {
 		return nil, fmt.Errorf("failed to update kubeone cluster status %s: %w", externalCluster.Name, err)
 	}
 
 	return newCluster, nil
 }
+
+func checkContainerRuntime(ctx context.Context,
+	externalCluster *kubermaticv1.ExternalCluster,
+	externalClusterProvider provider.ExternalClusterProvider,
+) (string, error) {
+	var containerRuntime string
+	nodes, err := externalClusterProvider.ListNodes(ctx, externalCluster)
+	if err != nil {
+		return "", fmt.Errorf("failed to list nodes: %w", err)
+	}
+	for _, node := range nodes.Items {
+		if _, ok := node.Labels[NodeControlPlaneLabel]; ok {
+			containerRuntimeVersion := node.Status.NodeInfo.ContainerRuntimeVersion
+			strSlice := strings.Split(containerRuntimeVersion, ":")
+			for _, v := range strSlice {
+				containerRuntime = v
```

why just return here

---

maybe return strSlice[0]

---

i just did this to avoid the case when the slice is empty, indexnotfound err.
but yes, we can return here.

---

otherwise, you will return an empty string. Is it OK?

---

we can return an err, in case of empty string.
wdyt?

---

err could be something like no container runtime found...

---

err or some default, I am not sure

---

the error will be better